### ISSUE TITLE
Redevelop a process for producing NBN export files

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1301]
When we last discussed this I think we decided that the best option would be to reinstate the existing method for producing export files, as an immediate solution while a new more automated process is being developed.

I'd like to make the reinstatement option a priority, as I am currently unable to generate new exports and a small queue of requests is building up.

See related issues at #1052 and #1235.

**Comment by johnvanbreda** [comment]
@kitenetter isn't the existing mechanism for generating export files still working? There are files generated yesterday as far as I can see.

I've just used the new Elasticsearch based process for exporting the Sawflies dataset and compared the output with the PostgreSQL based output from the old method. I found the following differences:
1. Currently, the Elasticsearch version of the scientificName field uses the accepted name whereas the PostgreSQL output will use the name as recorded, but only if this is a scientific name, otherwise if a common name is recorded it uses the accepted name.
2. Currently, if there is no identifier in the attributes for a record, Elasticsearch outputs nothing for the identifiedBy field whereas the old PostgreSQL output would fall back on the recorder name.

Shall I modify the Elasticsearch output to match the old output behaviour for these?

The only other difference I could spot was the Elasticsearch output includes a few more records as the method for filtering which names are included in Symphyta is more reliable.

**Comment by kitenetter** [comment]
The existing process is outputting files, but I am unable to trigger them when needed, and I am unable to add new filters, For example, "NBN - Auchenorrhyncha Recording Scheme" is not producing any files.

I think it is okay to output the current recommended name. That is what we do within the Explore pages and the verifier downloads isn't it, so it makes sense to align the NBN data with what the verifiers are seeing.

I think it is fine to leave identifedBy blank.

**Comment by johnvanbreda** [comment]
@kitenetter do you want to be able to run the exports ad-hoc, or should we set it up so they are automatically generated on a monthly basis?

@BirenRathod we need to think about how best to allow Martin to run the new exports. They are triggered by running a *.bat file, which uses PHP to run a script. The script needs to run from a machine that can see the Elasticsearch server (so inside the CEH network I think) and it will obviously need PHP installed.

**Comment by BirenRathod** [comment]
@johnvanbreda I Installed PHP. Not sure, which version do you need but I use 8.1. 

About running bat file, There is a folder created on C drive called indicia. You can create .bat file in there. I will give access to @kitenetter , so he can run that file when needed. Is that ok?

**Comment by johnvanbreda** [comment]
@BirenRathod which machine is the 'indicia' folder on?

**Comment by BirenRathod** [comment]
@johnvanbreda it is on elasticsearch server.

**Comment by johnvanbreda** [comment]
@BirenRathod thanks. I've copied over the files required now for 6 different datasets so @kitenetter can test this. There are 3 components to each dataset:
1. A *.bat file that triggers PHP to do the export (using PHP code in the ipt-dev folder).
2. A *.json file of the same name that contains the configuration, in particular the query used to obtain the records from Elasticsearch.
3. A folder containing a *.eml file and a metadata.xml file with the contents required by the DwCA output zip file.

So the scheduled tasks just need to run the command in the *.bat file and Martin can do the same to generate a file ad-hoc.

Martin, is there a particular dataset you'd like me to implement, perhaps with documentation on how I did it?

**Comment by BirenRathod** [comment]
@johnvanbreda I got a small request before giving permission to @kitenetter. Can you add export folder, so that each will be created in export folder. 
The idea is to keep those file & folsers generated by *.bat file, can keep separate from .bat files.

**Comment by kitenetter** [comment]
One of the new ones I wanted to set up was for leafhoppers and allies, which uses the filter "NBN - Auchenorrhyncha" and the auto-export "NBN - Auchenorrhyncha Recording Scheme".

Or if it makes more sense to test an export that we've done before I'd suggest "NBN - Heleomyzid Recording Scheme", which is a fairly small dataset that I can compare with the previous export.

Documentation would be much appreciated, including whether we are still using the same NBN filters (I assume we are) and how I set up the EML for any new exports.

**Comment by johnvanbreda** [comment]
@BirenRathod I've organised the exports, config files and metadata files into sub-folders.

@kitenetter I'll start by setting up the NBN - Heleomyzid Recording Scheme output. I've been obtaining the EML simply by copying it out of the EML field on the Export's edit tab on iRecord. I'm not sure how these were originally filled in though. Other than laboriously working through the EML specification to create a document manually, the only other approach I know of is to create a dataset on the IPT, then the view page for the dataset includes an EML document.

**Comment by kitenetter** [comment]
Thanks @johnvanbreda - results of comparison below. The EML data is created by me based on information from the dataset owner, and at the moment I copy it across into the NBN export set-up page on iRecord. If needed I can copy all of them somewhere else, but don't want to have to do that every time we're due to send an export.

Heleomyzid data file:

**1. Website agreements**
The new file has 28 records that are not in the old one. One of these is the result of a recently added record being included, all others are the result of data being included from websites that do not provide data for dataflow - in other words the new system is not respecting the website agreements.

**2. Record-level licences**
The old file has two records that are not in the new one. The two records IDs are 23419652 and 23419653, both originating from "Nature Counts - Nottinghamshire WT | General records", which does have a dataflow sharing agreement. They are the only two records in this dataset that have a record-level licence ("CC BY-NC"), so my assumption is that record-level licences are not being correctly handled by the new process.

**3. Lat/long reversed**
The latitude and longitude data has been transposed in the new version (i.e. the column headings are in the same order as before, but the lat column contains the long data, and vice versa).

**4. Date range within single month**
In the old system, date ranges within a single month were shown as:
- 1982-09-13/22
- 2019-12-14/28

The new system currently has them as:
- 1982-09-13/1982-09-22
- 2019-12-14/2019-12-28

I don't see why NBN Atlas couldn't use the new format, but I will check with them to find out.

**5. Other**
Where a record is marked as sensitive we are outputting a null location name in both old and new systems. This is probably fine, but I suppose there is an argument for making it explicit that the name has been hidden, to distinguish from those records where the name was never entered.

For completeness, the other difference is that, for cases where "Identified by" is null, the new system keeps it as null, whereas the old system was populating it with the recorder name. I think the new system is the better option, as it more truthfully reflects what was originally recorded.

**Comment by johnvanbreda** [comment]
I've sent a new file. Some responses:
1 & 2 - these were caused by me using the reporting agreement to generate the Elasticsearch query, rather than the data flow agreement. I've now corrected this.
3. Lat/long reversal fixed.
4. OK - let me know if the date range format generation needs improvement.
5. Let me know if the location output needs changing for sensitive records, or the identified behaviour.

**Comment by kitenetter** [comment]
For no. 4 Sophie confirms that the new date range format should be fine.

**Comment by kitenetter** [comment]
@johnvanbreda I will need to send data files to NBN again at the end of next week (i.e. by 1 July). Are we likely to be able to use the new system for this? If so is it possible to set up new filters for the datasets that we need to add to the list?

**Comment by kitenetter** [comment]
Thanks Biren, I can now access that folder.

What happens next - what do I need to do to ensure that existing exports are generated, and to set up the new ones?

**Comment by BirenRathod** [comment]
@johnvanbreda will explain it you.

**Comment by johnvanbreda** [comment]
@kitenetter for the issue about the location name for sensitive records, please can you raise this as a separate issue so the scope doesn't creep (as the old export behaves the same way as the new). Also it might be that you want this to apply across the board, e.g. when looking at records in the Record Details page, or Explore page - so not just an export issue.

At the moment I don't think Biren has automated the export batch files, but the exports can be run by going to the c:\indicia folder and running a *.bat file - this will need to be run on the server itself, not just via a shared network drive though. 

I just created the missing configuration for the Ladybird dataset and generated the file to see how long it would take me. It took about 15 minutes to set everything up. BUT, the export file generated was about twice the size of the last one generated using the old method. So, I had to compare the datasets to look for the additional records and see if I could spot the reason why. There were 2 reasons, first, quite a few additional records had been verified. The second reason was that the new Elasticsearch way of filtering Coccinellidae is more effective than the old PostgreSQL way as data from flat lists get included, as long as the NBN Key is correct. So quite a lot of additional data gets included due to a lot of older records being linked to the flat iRecord Ladybirds list.

Given this information about the time required to create the config for each export and the need to check the outputs to at least understand the differences, do you want me to try and proceed with converting the other datasets to ES? Or maybe I could just do a handful, perhaps focusing on the big ones you had problems with before?

**Comment by kitenetter** [comment]
Location name for sensitive records: this is not a priority so I will park this for now.

One priority is to create a file for the leafhopper scheme. This was set up under the old system but I don't think it ever generated a file. The set-up page is:
https://irecord.org.uk/nbn-auchenorrhyncha-recording-scheme

For the rest, I'm not sure where the balance lies between @johnvanbreda spending time on setting up the new process or me spending time running the pgAdmin queries for datasets that fail to generate correctly. The largest files, which most frequently fail, are:
- Vascular plant records verified via iRecord
- Bee, wasp and ant records verified via iRecord
- National Mammal Atlas Project, online recording
- All taxa records for Leicestershire and Rutland

As far as I'm aware I don't have access to the server to run the batch files.

In the folder for the old export process (\\192.171.199.203) there are currently only 15 data files out of the 60 or so that we normally send - @BirenRathod is there a reason why only a few are available at the moment?

**Comment by BirenRathod** [comment]
@kitenetter Please don't get confused with new porcess and old expoert folder. There is no connection between them. 
The old export folder have got 15 data files becuase I haven't copied in them yet. I have now copied them manually.

**Comment by johnvanbreda** [comment]
@kitenetter The auchenorrhyncha export doesn't seem to be on the old copy of iRecord, so maybe it was only set up on Drupal 9. I've just generated it using the new technique so you can find it on the ES server.

@BirenRathod am I correct in thinking that the exports on the new Drupal 9 iRecord on Pantheon don't actually get run? And that the 2 ways we can generate a file is via the old PostgreSQL method using the copy of iRecord on the Drupal live server, or using the Elasticsearch server approach? If so where did you just copy all the files from?

**Comment by johnvanbreda** [comment]
@kitenetter I will set up new-style exports for the 4 largest file examples you give above.

**Comment by BirenRathod** [comment]
@johnvanbreda yes, that is right. there is no exports running from new ireCord Drupal 9 site on Pantheon.

The data file which I copied on old exports are from old irecord Drupal 7 site. These are from BRC server.

**Comment by kitenetter** [comment]
Thanks @johnvanbreda and @BirenRathod 

Biren - please can you trigger a new export file for "Diptera records from iRecord for families not covered by a recording scheme" (which is generated using auto-export "NBN_Diptera_non_scheme"). I had to change the filter today.

**Comment by kitenetter** [comment]
@BirenRathod Just realised that the filter I changed was on the Drupal 9 site, so I need to edit the filter on the Drupal 7 site. Can you remind me of the URL to get to the old site?

**Comment by kitenetter** [comment]
The filter I edited in D9 was:
https://irecord.org.uk/filters/admin-edit?filters_user_id=6811

That uses a fairly long list of families in the taxon part of the filter. I had to remove three families: Coelopidae, Heterocheila and Helcomyzidae.

I was assuming that this edit would need to be done on the D7 site in order to generate the correct file, but if I go to http://192.171.199.201/filters/admin-list it looks like the edit I did in D9 is already showing. Which makes sense, because the filters are stored in the warehouse aren't they. (Sorry, I am being very slow to understand how this all fits together!)

I have now re-triggered the export from 192.171.199.201/exports so hopefully that will update the export file later today, and it can then be copied over tomorrow.

**Comment by BirenRathod** [comment]
Thanks, @kitenetter.
I just checked and nothing in there, so I will copy it tomorrow.

**Comment by johnvanbreda** [comment]
@kitenetter the 4 largest export examples you gave above have all now been configured and generated using Elasticsearch.

**Comment by kitenetter** [comment]
@BirenRathod in the Exports folder the file "Diptera records from iRecord for families not covered by a recording scheme.zip" is dated 03/06/2022. Please can you do a final check to see if there is a more recent version available?

**Comment by BirenRathod** [comment]
@kitenetter I have copied files including Diptera one. I haven't checked the list but it should have all of those you mentioned.

**Comment by kitenetter** [comment]
I think we do now have a functioning process for generating the export files, but I'm still unclear about who needs to do what to generate a complete list of files each time we do an export. Questions:
1. Are we still using the NBN export filter settings that are stored in the warehouse (so that if I edit the filters the correct data will come through to the export file)?
2. Are all the exports now in the new format or do we still have a hybrid system?
3. I am not clear on who has to do what to make the new system work - my guess is that John has do the initial set up for the new process, but I'm not sure how it gets triggered each month, and where the data files end up.
4. What's the procedure for adding a new export?

**Comment by kitenetter** [comment]
As things stand, new export filters have to be set up by John. When exports are needed, the batch process needs to be triggered, by John (or Biren?). If a warehouse filter changes for one of the new exports, the export json file needs to be updated (that file is accessible from the indicia folder shown below, and can in principle be edited by anyone with access to the folder).

The new process produces separate data and metadata files, which need to be combined to send to NBN Atlas.

In the longer term it would be good to improve this process to match the functionality we used to have, e.g. so that filter changes are picked up automatically, and the batch files can be run automatically, and the source eml files can be edited, and the data and metadata files can be combined into one zip file.

File locations for reference;
- new exports: indicia (\\192.171.199.233)
- old exports: exports (\\192.171.199.203)

Closing this issue, will add new issues when specific edits or updates are needed.
